File: lib_compare/debug_kalman_variance.py
"""
Debug script to trace kalman_variance calculation step-by-step in Python
to compare with TypeScript implementation.
"""

import sys
import os
from datetime import datetime, timezone

# Add python_lib to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python_lib', 'src'))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'be_implementation_service', 'src'))

from weight_processor_lib.core.processing.processor import process_measurement
from weight_processor_lib.core.database.memory_store import InMemoryStore
from aws.config.config_manager import ConfigManager

def debug_kalman_variance():
    print('=' * 80)
    print('DEBUG: Kalman Variance Calculation (Python)')
    print('=' * 80)

    # Load config
    config_path = os.path.join(os.path.dirname(__file__), '..', 'python_lib', 'config.toml')
    config = ConfigManager.load_config(config_path)

    # Create store
    store = InMemoryStore()
    user_id = 'debug-user:debug-device'

    # Fixed timestamps for reproducibility (match TypeScript)
    timestamp1 = datetime(2025, 11, 10, 10, 0, 0, tzinfo=timezone.utc)
    timestamp2 = datetime(2025, 11, 11, 10, 0, 0, tzinfo=timezone.utc)  # 1 day later

    print('\n' + '=' * 80)
    print('MEASUREMENT 1: Initialize Kalman filter')
    print('=' * 80)
    print(f'Weight: 70.0 kg')
    print(f'Timestamp: {timestamp1.isoformat()}')
    print(f'Source: withings')

    result1 = process_measurement(
        user_id=user_id,
        weight=70.0,
        timestamp=timestamp1,
        source='withings',
        config=config,
        unit='kg',
        db=store,
        user_height_m=None
    )

    print('\nResult 1:')
    print(f'  filtered_weight: {result1.get("filtered_weight")}')
    print(f'  trend: {result1.get("trend")}')
    print(f'  kalman_variance: {result1.get("kalman_variance")}')
    print(f'  kalman_confidence_upper: {result1.get("kalman_confidence_upper")}')
    print(f'  kalman_confidence_lower: {result1.get("kalman_confidence_lower")}')

    # Get state after first measurement
    state1 = store.get_state(user_id)
    print('\nState after measurement 1:')
    print(f'  last_state shape: {state1["last_state"].shape}')
    print(f'  last_state[0]: {state1["last_state"][0]}')
    print(f'  last_state[1]: {state1["last_state"][1]}')
    print(f'  last_covariance shape: {state1["last_covariance"].shape}')
    print(f'  last_covariance[0]: ')
    print(f'    {state1["last_covariance"][0]}')
    print(f'  last_covariance[1]: ')
    print(f'    {state1["last_covariance"][1]}')
    print(f'  last_covariance[0][0,0]: {state1["last_covariance"][0][0,0]}')
    print(f'  last_covariance[1][0,0]: {state1["last_covariance"][1][0,0]}')

    if 'kalman_params' in state1:
        print('\nKalman params:')
        print(f'  observation_covariance: {state1["kalman_params"]["observation_covariance"]}')
        print(f'  transition_covariance: {state1["kalman_params"]["transition_covariance"]}')

    print('\n' + '=' * 80)
    print('MEASUREMENT 2: Update with prediction step (1 day later)')
    print('=' * 80)
    print(f'Weight: 70.1 kg')
    print(f'Timestamp: {timestamp2.isoformat()}')
    print(f'Time delta: 1 day')
    print(f'Source: withings')

    # Enable verbose logging
    os.environ['VERBOSE_LOGGING'] = 'true'

    result2 = process_measurement(
        user_id=user_id,
        weight=70.1,
        timestamp=timestamp2,
        source='withings',
        config=config,
        unit='kg',
        db=store,
        user_height_m=None
    )

    os.environ['VERBOSE_LOGGING'] = 'false'

    print('\nResult 2:')
    print(f'  filtered_weight: {result2.get("filtered_weight")}')
    print(f'  trend: {result2.get("trend")}')
    print(f'  kalman_variance: {result2.get("kalman_variance")}')
    print(f'  innovation: {result2.get("innovation")}')
    print(f'  normalized_innovation: {result2.get("normalized_innovation")}')
    print(f'  kalman_confidence_upper: {result2.get("kalman_confidence_upper")}')
    print(f'  kalman_confidence_lower: {result2.get("kalman_confidence_lower")}')

    # Get final state
    state2 = store.get_state(user_id)
    print('\nState after measurement 2:')
    print(f'  last_state[0]: {state2["last_state"][0]}')
    print(f'  last_state[1]: {state2["last_state"][1]}')
    print(f'  last_covariance[0]: ')
    print(f'    {state2["last_covariance"][0]}')
    print(f'  last_covariance[1]: ')
    print(f'    {state2["last_covariance"][1]}')
    print(f'  last_covariance[0][0,0]: {state2["last_covariance"][0][0,0]}')
    print(f'  last_covariance[1][0,0]: {state2["last_covariance"][1][0,0]}')

    print('\n' + '=' * 80)
    print('SUMMARY')
    print('=' * 80)
    print(f'Python kalman_variance (measurement 2): {result2.get("kalman_variance")}')
    print(f'Expected TypeScript value: 2.379729588093491')
    if result2.get('kalman_variance'):
        diff = result2.get('kalman_variance') - 2.379729588093491
        rel_err = (abs(diff) / result2.get('kalman_variance')) * 100
        print(f'Difference: {diff:.6f}')
        print(f'Relative error: {rel_err:.2f}%')

if __name__ == '__main__':
    debug_kalman_variance()

File: lib_compare/debug_kalman_variance.ts
/**
 * Debug script to trace kalman_variance calculation step-by-step
 * to find exact divergence between Python and TypeScript
 */

import { processMeasurement } from '../typescript_lib/src/weight-processor-lib/core/processing/processor';
import { InMemoryStore } from '../typescript_lib/src/weight-processor-lib/core/database/memory_store';

async function debugKalmanVariance() {
  console.log('='.repeat(80));
  console.log('DEBUG: Kalman Variance Calculation');
  console.log('='.repeat(80));

  // Load config
  const configPath = `${import.meta.dir}/../typescript_lib/config.json`;
  const config = await Bun.file(configPath).json();

  // Create store
  const store = new InMemoryStore();
  const userId = 'debug-user:debug-device';

  // Fixed timestamps for reproducibility
  const timestamp1 = new Date('2025-11-10T10:00:00.000Z');
  const timestamp2 = new Date(timestamp1.getTime() + 86400000); // 1 day later

  console.log('\n' + '='.repeat(80));
  console.log('MEASUREMENT 1: Initialize Kalman filter');
  console.log('='.repeat(80));
  console.log(`Weight: 70.0 kg`);
  console.log(`Timestamp: ${timestamp1.toISOString()}`);
  console.log(`Source: withings`);

  const result1 = await processMeasurement(
    userId,
    70.0,
    timestamp1,
    'withings',
    config,
    'kg',
    store,
    null
  );

  console.log('\nResult 1:');
  console.log(`  filtered_weight: ${result1.filtered_weight}`);
  console.log(`  trend: ${result1.trend}`);
  console.log(`  kalman_variance: ${result1.kalman_variance}`);
  console.log(`  kalman_confidence_upper: ${result1.kalman_confidence_upper}`);
  console.log(`  kalman_confidence_lower: ${result1.kalman_confidence_lower}`);

  // Get state after first measurement
  const state1 = await store.getState(userId);
  console.log('\nState after measurement 1:');
  console.log(`  last_state shape: ${state1?.last_state?.length} x ${state1?.last_state?.[0]?.rows}x${state1?.last_state?.[0]?.columns}`);
  console.log(`  last_state[0]: [${state1?.last_state?.[0]?.to1DArray()}]`);
  console.log(`  last_state[1]: [${state1?.last_state?.[1]?.to1DArray()}]`);
  console.log(`  last_covariance shape: ${state1?.last_covariance?.length}`);
  console.log(`  last_covariance[0][0,0]: ${state1?.last_covariance?.[0]?.get(0, 0)}`);
  console.log(`  last_covariance[1][0,0]: ${state1?.last_covariance?.[1]?.get(0, 0)}`);

  if (state1?.kalman_params) {
    console.log('\nKalman params:');
    console.log(`  observation_covariance: ${JSON.stringify(state1.kalman_params.observation_covariance)}`);
    console.log(`  transition_covariance: ${JSON.stringify(state1.kalman_params.transition_covariance)}`);
  }

  console.log('\n' + '='.repeat(80));
  console.log('MEASUREMENT 2: Update with prediction step (1 day later)');
  console.log('='.repeat(80));
  console.log(`Weight: 70.1 kg`);
  console.log(`Timestamp: ${timestamp2.toISOString()}`);
  console.log(`Time delta: 1 day`);
  console.log(`Source: withings`);

  // Enable verbose logging for this measurement
  process.env.VERBOSE_LOGGING = 'true';

  const result2 = await processMeasurement(
    userId,
    70.1,
    timestamp2,
    'withings',
    config,
    'kg',
    store,
    null
  );

  process.env.VERBOSE_LOGGING = 'false';

  console.log('\nResult 2:');
  console.log(`  filtered_weight: ${result2.filtered_weight}`);
  console.log(`  trend: ${result2.trend}`);
  console.log(`  kalman_variance: ${result2.kalman_variance}`);
  console.log(`  innovation: ${result2.innovation}`);
  console.log(`  normalized_innovation: ${result2.normalized_innovation}`);
  console.log(`  kalman_confidence_upper: ${result2.kalman_confidence_upper}`);
  console.log(`  kalman_confidence_lower: ${result2.kalman_confidence_lower}`);

  // Get final state
  const state2 = await store.getState(userId);
  console.log('\nState after measurement 2:');
  console.log(`  last_state[0]: [${state2?.last_state?.[0]?.to1DArray()}]`);
  console.log(`  last_state[1]: [${state2?.last_state?.[1]?.to1DArray()}]`);
  console.log(`  last_covariance[0][0,0]: ${state2?.last_covariance?.[0]?.get(0, 0)}`);
  console.log(`  last_covariance[1][0,0]: ${state2?.last_covariance?.[1]?.get(0, 0)}`);

  console.log('\n' + '='.repeat(80));
  console.log('SUMMARY');
  console.log('='.repeat(80));
  console.log(`TypeScript kalman_variance (measurement 2): ${result2.kalman_variance}`);
  console.log(`Expected Python value: 4.00252950373697`);
  console.log(`Difference: ${result2.kalman_variance ? (4.00252950373697 - result2.kalman_variance).toFixed(6) : 'N/A'}`);
  console.log(`Relative error: ${result2.kalman_variance ? ((Math.abs(4.00252950373697 - result2.kalman_variance) / 4.00252950373697) * 100).toFixed(2) : 'N/A'}%`);
}

// Run the debug
debugKalmanVariance().catch(console.error);
